Ship_Game/Universe/SolarBodies/DysonSwarm.cs

```diff
@@ -0,0 +1,237 @@
+                SwarmSatProductionCost *= 3;
+
+            Init();
+        }
+
+        [StarDataConstructor]
+        public DysonSwarm() { }
+
+        void Init()
+        {
+            AddControllerPositions(24, 5000);
+            AddControllerPositions(12, 3750);
+            AddControllerPositions(9,  2500);
+            AddControllerPositions(5,  1250);  // total of 50
+
+            void AddControllerPositions(int numPerRing, int distance)
+            {
+                float degrees = 360 / numPerRing;
+                for (int i = 0; i < numPerRing; i++)
+                {
+                    var offset = MathExt.PointOnCircle(i * degrees, distance);
+                    SwarmControllers.Add(System.Position + offset, null);
+                }
+            }
+
+            foreach (Planet p in System.PlanetList) 
+            {
+                if (p.Owner == Owner) 
+                    p.SetDysonSwarmWeapon(loadWeapon: true);
+            }
+        }
+
+        public void Update() // Once per turn or when a new Dyson Swarm Sat is deployed
+        {
+            int count = 0;
+            foreach (KeyValuePair<Vector2, Ship> item in SwarmControllers)
+            {
+                Ship swarmController = item.Value;
+                if (swarmController != null)
+                {
+                    if      (!swarmController.Active)          SwarmControllers[item.Key] = null;
+                    else if (swarmController.Loyalty != Owner) swarmController.AI.OrderScuttleShip();
+                    else                                       count++;
+                }
+            }
+
+            UpdateMaxOverclock();
+            ControllerCompletion = count / (float)TotalSwarmControllers;
+            int desiredOverclock = OverclockEnabled ? MaxOverclock : 0;
+            CurrentOverclock += (desiredOverclock - CurrentOverclock).Clamped(-1, 1);
+            float completionLimit = ControllerCompletion.UpperBound(SwarmCompletion);
+            CurrentOverclock = CurrentOverclock.UpperBound((int)(completionLimit * MaxOverclock));
+            if (NeedDysonRingsChange)
+                LoadDysonRings();
+        }
+
+        public void UpdateMaxOverclock()
+        {
+            MaxOverclock = Owner.data.Traits.DysonSwarmMaxOverclock;
+        }
+
+        public void KillSwarm()
+        {
+            float scuttleTimer = 1;
+            foreach (Ship swarmController in SwarmControllers.Values)
+            {
+                if (swarmController?.Active == true) 
+                {
+                    swarmController.ScuttleTimer = scuttleTimer;
+                    scuttleTimer += 0.25f;
+                }
+            }
+
+            NumSwarmSats = 0;
+        }
+
+        public void DeploySwarmSat()
+        {
+            NumSwarmSats = (NumSwarmSats + 1).UpperBound(RequiredSwarmSats);
+        }
+      
+        void LoadDysonRings()
+        {
+            int neededRings = (int)(SwarmCompletion * 100) / 10;
```

Comment:
wouldn't this line turn int back to float? Maybe use some round function? Or do casting to int in the end?

Reply:
SwarmCompletion is float. It gets converted to 1-100 and to Int. Int divided by int gives a floored int.
so if the result is 95/10 = 9.5 which is int 9. This is what i want here.